### PR TITLE
[SHELL32][SHELLEXT] Fix use of SHFileOperation results and improve log on failure / fix log spam

### PR DESCRIPTION
**Pull request by KyleKatarn31**
## Purpose

- We are currently not logging failures of SHFileOperation in a systematic and consistent way
- Some ERR() are unduly triggered for normal operations
- We use GetLastError() while MSDN states : "Do not use GetLastError with the return values of this function."
- Result sometime wrongly used as HRESULT

## Proposed changes

- Systematic design pattern for SHFileOperation logging
- Explicit use of SHFileOperationW
- Replace ERR() by TRACE() for normal operations
- ERR() shows return code instead of GetLastError as per MSDN
- Result never used as HRESULT

### CODE REVIEW

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp [hunk 1]
```diff
@@ -157,7 +157,9 @@ CMyDocsDropHandler::Drop(IDataObject *pDataObject, DWORD dwKeyState,
     fileop.pFrom = pszzSrcList;
     fileop.pTo = szzDir;
     fileop.fFlags = FOF_ALLOWUNDO | FOF_FILESONLY | FOF_MULTIDESTFILES | FOF_NOCONFIRMMKDIR;
-    SHFileOperationW(&fileop);
+    hr = SHFileOperationW(&fileop);
```
Comment (learn-more) [inline dll/shellext/mydocs/CMyDocsDropHandler.cpp:160]: if it does not produce an HRESULT, dont call it hr.

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp [hunk 2]
```diff
@@ -157,7 +157,9 @@ CMyDocsDropHandler::Drop(IDataObject *pDataObject, DWORD dwKeyState,
...
+    if (hr)
+        ERR("SHFileOperationW failed with 0x%x\n", GetLastError());
```
Comment (learn-more) [inline dll/shellext/mydocs/CMyDocsDropHandler.cpp:162]: msdn states:
> Do not use GetLastError with the return values of this function.

**Review by learn-more — CHANGES_REQUESTED** [review]
Why do all these cases need to log on failure?
Is the user supposed to read the debug log to figure out what is wrong?

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp (hunk 2 above)
Comment (KyleKatarn31) [reply dll/shellext/mydocs/CMyDocsDropHandler.cpp:162]: Good point, I did not notice... but we already do it elsewhere in ROS for this function... i'll fix !

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp (hunk 1 above)
Comment (KyleKatarn31) [reply dll/shellext/mydocs/CMyDocsDropHandler.cpp:160]: hr is already the type used for the return of the calling function ... which misses to fail when SHFileOperationW fails...

**Comment by KyleKatarn31** [comment]
> Why do all these cases need to log on failure?
> Is the user supposed to read the debug log to figure out what is wrong?

I've facing cases where such failures only produce ERR traces without effect to users (no error dialog box at explorer level). This is something we already discussed and i'll look at this, but this systematic logging will help to ensure full coverage.

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp (hunk 1 above)
Comment (HBelusca) [reply dll/shellext/mydocs/CMyDocsDropHandler.cpp:160]: SHFileOperationW does not return errors in an HRESULT format (and neither in the usual Win32 last error format). So as @learn-more hinted at, using hr (that is also used to return something from the calling function) is a mistake.

**Review by HBelusca — CHANGES_REQUESTED** [review]

**Comment by KyleKatarn31** [comment]
OK, i'll change "hr" to int and will fix where hr was already used in the current code.

File: dll/win32/shell32/shlfileop.cpp [hunk 3]
```diff
@@ -1027,6 +1027,8 @@ int WINAPI SHFileOperationA(LPSHFILEOPSTRUCTA lpFileOp)
 
     // Call the actual function
     retCode = SHFileOperationW(&nFileOp);
+    if (retCode)
```
Comment (SergeGautherie) [inline dll/win32/shell32/shlfileop.cpp:1030]: Missing braces.

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp [hunk 4]
```diff
@@ -157,7 +157,12 @@ CMyDocsDropHandler::Drop(IDataObject *pDataObject, DWORD dwKeyState,
     fileop.pFrom = pszzSrcList;
     fileop.pTo = szzDir;
     fileop.fFlags = FOF_ALLOWUNDO | FOF_FILESONLY | FOF_MULTIDESTFILES | FOF_NOCONFIRMMKDIR;
-    SHFileOperationW(&fileop);
+    int res = SHFileOperationW(&fileop);
+    if (res)
```
Comment (SergeGautherie) [inline dll/shellext/mydocs/CMyDocsDropHandler.cpp:161]: Maybe, choose between `if (res)` and `if (res != 0)`, then use the same pattern everywhere?

File: dll/win32/shell32/droptargets/CFSDropTarget.cpp [hunk 5]
```diff
@@ -92,7 +92,10 @@ HRESULT CFSDropTarget::_CopyItems(IShellFolder * pSFFrom, UINT cidl,
     HeapFree(GetProcessHeap(), 0, pszSrcList);
 
     if (res)
+    {
+        ERR("SHFileOperationW failed with 0x%x\n", res);
         return E_FAIL;
+    }
     else
         return S_OK;
```
Comment (SergeGautherie) [inline dll/win32/shell32/droptargets/CFSDropTarget.cpp:100]: Remove `else`, or add braces.

File: dll/win32/shell32/shellrecyclebin/recyclebin_v5.c [hunk 6]
```diff
@@ -503,8 +503,10 @@ RecycleBin5_RecycleBin5_Restore(
             op.pFrom = pDeletedFileName;
             op.pTo = pDeletedFile->FileNameW;
 
-            if (!SHFileOperationW(&op))
+            int res = SHFileOperationW(&op);
+            if (!res)
```
Comment (SergeGautherie) [inline dll/win32/shell32/shellrecyclebin/recyclebin_v5.c:507]: Existing bug!?
```suggestion
            if (res)
```

File: dll/win32/shell32/shellrecyclebin/recyclebin_v5.c [hunk 7]
```diff
@@ -503,8 +503,10 @@ RecycleBin5_RecycleBin5_Restore(
...
             {
+                ERR("SHFileOperationW failed with 0x%x\n", res);
                 UnmapViewOfFile(pHeader);
                 return HRESULT_FROM_WIN32(GetLastError());
```
Comment (SergeGautherie) [inline dll/win32/shell32/shellrecyclebin/recyclebin_v5.c:511]: Existing bug!?
```suggestion
                return E_FAIL;
```

**Review by SergeGautherie — CHANGES_REQUESTED** [review]

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp (hunk 4 above)
Comment (KyleKatarn31) [reply dll/shellext/mydocs/CMyDocsDropHandler.cpp:161]: OK

File: dll/win32/shell32/shellrecyclebin/recyclebin_v5.c (hunk 7 above)
Comment (KyleKatarn31) [reply dll/win32/shell32/shellrecyclebin/recyclebin_v5.c]: Yes.

**Comment by KyleKatarn31** [comment]
@HBelusca @SergeGautherie : For final review after last round of comments
@binarymaster : I'd suggest bugfix rather than "trivial" following bugs highlighted by @SergeGautherie and @learn-more on incorrect use of GetLastError with this API

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp (hunk 4 above)
Comment (SergeGautherie) [inline dll/shellext/mydocs/CMyDocsDropHandler.cpp:161]: Replace `return hr;` with this block and `return S_OK;` too.

File: dll/win32/shell32/droptargets/CFSDropTarget.cpp [hunk 8]
```diff
@@ -680,7 +682,13 @@ HRESULT CFSDropTarget::_DoDrop(IDataObject *pDataObject,
             op.hwnd = m_hwndSite;
             op.wFunc = bCopy ? FO_COPY : FO_MOVE;
             op.fFlags = FOF_ALLOWUNDO | FOF_NOCONFIRMMKDIR;
-            hr = SHFileOperationW(&op);
+            int res = SHFileOperationW(&op);
+            if (res)
```
Comment (SergeGautherie) [inline dll/win32/shell32/droptargets/CFSDropTarget.cpp:686]: Replace `return hr;` with this block and `return S_OK;` too.

While here, `hr` at line 490/492 should be used or removed...

File: dll/win32/shell32/droptargets/CRecyclerDropTarget.cpp [hunk 9]
```diff
@@ -58,9 +58,10 @@ class CRecyclerDropTarget :
                 FileOp.fFlags = FOF_ALLOWUNDO;
             TRACE("Deleting file (just the first) = %s, allowundo: %d\n", debugstr_w(FileOp.pFrom), (FileOp.fFlags == FOF_ALLOWUNDO));
 
-            if (SHFileOperationW(&FileOp) != 0)
+            int res = SHFileOperationW(&FileOp);
+            if (res)
```
Comment (SergeGautherie) [inline dll/win32/shell32/droptargets/CRecyclerDropTarget.cpp:62]: Replace `return hr;` with this block and `return S_OK;` too.

**Review by SergeGautherie — CHANGES_REQUESTED** [review]

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp (hunk 4 above)
Comment (HBelusca) [reply dll/shellext/mydocs/CMyDocsDropHandler.cpp:161]: Wth are you talking about @SergeGautherie ?

File: dll/win32/shell32/droptargets/CFSDropTarget.cpp (hunk 8 above)
Comment (KyleKatarn31) [reply dll/win32/shell32/droptargets/CFSDropTarget.cpp:686]: OK to use it.

File: dll/shellext/mydocs/CMyDocsDropHandler.cpp (hunk 4 above)
Comment (KyleKatarn31) [reply dll/shellext/mydocs/CMyDocsDropHandler.cpp:161]: Not understood. Marked as resolved unless clarified.

**Comment by KyleKatarn31** [comment]
@HBelusca : ready to go ?

**Review by HBelusca — APPROVED** [review]

**Pull request merged**